src/outputs/parthenon_opmd.cpp

```diff
@@ -0,0 +1,672 @@
+  WriteAllParamsOfMultipleTypes<PARTHENON_ATTR_VALID_VEC_TYPES(T)>(params, prefix, it);
+}
+
+void WriteAllParams(const Params &params, const std::string &pkg_name,
+                    openPMD::Iteration *it) {
+  using OpenPMDUtils::delim;
+  const std::string prefix = "Params" + delim + pkg_name;
+  // check why this (vector of bool) doesn't work
+  // WriteAllParams<bool>(params, prefix, it);
+  WriteAllParamsOfType<bool>(params, prefix, it);
+  WriteAllParams<int32_t>(params, prefix, it);
+  WriteAllParams<int64_t>(params, prefix, it);
+  WriteAllParams<uint32_t>(params, prefix, it);
+  WriteAllParams<uint64_t>(params, prefix, it);
+  WriteAllParams<float>(params, prefix, it);
+  WriteAllParams<double>(params, prefix, it);
+  WriteAllParams<std::string>(params, prefix, it);
+}
+
+template <typename T>
+void WriteSwarmVar(const SwarmInfo &swinfo, openPMD::ParticleSpecies swm,
+                   openPMD::Iteration it) {
+  auto &vars_of_type_T = std::get<SwarmInfo::MapToVarVec<T>>(swinfo.vars);
+  for (const auto &[vname, swmvarvec] : vars_of_type_T) {
+    const auto &vinfo = swinfo.var_info.at(vname);
+    auto host_data = swinfo.FillHostBuffer<T>(vname, swmvarvec);
+
+    auto const dataset = openPMD::Dataset(openPMD::determineDatatype(host_data.data()),
+                                          {swinfo.global_count});
+    // TODO(pgrete) ask OpenPMD group if this is the right approach of if our non-scalar
+    // partices should be a multi-D `dataset` if is scalar
+    if (vinfo.tensor_rank == 0) {
+      // special sauce to align "positions" with standard
+      std::string particle_record;
+      std::string particle_record_component;
+      if (vname == swarm_position::x::name()) {
+        particle_record = "position";
+        particle_record_component = "x";
+      } else if (vname == swarm_position::y::name()) {
+        particle_record = "position";
+        particle_record_component = "y";
+      } else if (vname == swarm_position::z::name()) {
+        particle_record = "position";
+        particle_record_component = "z";
+      } else {
+        particle_record = vname;
+        particle_record_component = openPMD::MeshRecordComponent::SCALAR;
+      }
+      openPMD::RecordComponent rc = swm[particle_record][particle_record_component];
+      rc.resetDataset(dataset);
+      // only write if there's sth to write (otherwise the host_data nullptr is caught)
+      if (swinfo.count_on_rank != 0) {
+        rc.storeChunk(host_data, {swinfo.global_offset}, {host_data.size()});
+      }
+
+      // if positional, add offsets
+      if (particle_record_component != openPMD::MeshRecordComponent::SCALAR) {
+        auto rc_offset = swm["positionOffset"][particle_record_component];
+        rc_offset.resetDataset(dataset);
+        rc_offset.makeConstant(0.0);
+      }
+
+      // else flatten components
+    } else {
+      for (auto n = 0; n < vinfo.nvar; n++) {
+        openPMD::RecordComponent rc = swm[vname][std::to_string(n)];
```

Comment:
I think there's an issue here. According to the OpenPMD spec (https://github.com/openPMD/openPMD-standard/blob/latest/STANDARD.md#records-for-each-particle-species), it's required to have `position/x` etc., defined for each particle species.

If there's isn't a var with the name "position", then I think the files won't be compliant (and as a result, I currently can't read AthenaPK tracer particles with my VisIt plugin).

Reply:
Ah, nevermind. I see this is correctly handled. I was seeing a different issue.